N-argument convolution (redone)

HDictus opened the pull request:
I messed up the N-args branch pretty badly and couldn't rebase to overlap-save, so I manually moved the changes to a new branch based on up-to-date overlap-save. Commits should be cleaner this way anyway.
This PR isn't for merging, but to have a clear diff until overlap-save is merged.

galenlynch committed:
Overlap-save convolution

I have added an overlap-save algorithm to convolution, which is substantially
faster (~2x) for large convolution problems. It is also substantially faster
than other packages like ImageFiltering.jl. For smaller convolutions, it is
about the same or a little bit slower than standard fft convolution. The cutoff
between overlap-save and simple fft convolution is pretty arbitrary at the
moment, so it would be nice to add some benchmarking to allow DSP.jl to choose
between the two algorithms depending on the size of the input. It would also be
nice to add spatial/time domain filtering for small convolutions, as it would
almost certainly be faster than fft convolutions. This is relevant to #224.

galenlynch committed:
Explain _conv_kern_os_edge! parameters

galenlynch committed:
Lots of sign posting

galenlynch committed:
Move edge_dims_arr declaration closer to where it's used

galenlynch committed:
Use fewer loops in perimeter iteration

galenlynch committed:
Speed up testing of conv by using fewer combinations of sizes

galenlynch committed:
Simplify Val construction

galenlynch committed:
Mark _conv_kern_os as unsafe

galenlynch committed:
Do not allow optimalfftfiltlength to change the order of inputs

galenlynch committed:
Turn off inbounds access in overlap-save, for now

galenlynch committed:
Remove unnecessary anonymous function in map call

galenlynch committed:
Remove redundant test

HDictus committed:
add tests for multi-argument convolution

HDictus committed:
implement N-argument convolution

performs overlap-save on the first argument if it is large
enough to warrant it. Performs fft convolution of all others.

galenlynch commented:
ok thanks!

galenlynch commented on src/dspbase.jl:
```diff
@@ -432,15 +437,38 @@ function unsafe_conv_kern_os_edge!(
                     out_start .+ data_offset .+ save_blocksize .- 1,
                     out_stop
                 )
+                # print("data offset", data_offset, "\n")
```
Looks like you don't know about [`@show`](https://docs.julialang.org/en/v1/base/base/#Base.@show), it's great!

HDictus replied:
I did not, thanks!
Also, whoops.

HDictus committed:
remove print statements

galenlynch committed:
Reduce diff noise

galenlynch committed:
sv is already passed as a parameter to this function

galenlynch commented:
Can you allow me to edit the PR?

HDictus commented:
'Allow edits from maintainers' is checked off, so you should be able to. Unless you mean something else?

galenlynch commented:
No that's what I was asking about: I've just never done that before so it took me a little to figure out how to do it.

I made some commits to your PR to reduce what shows up in the diff, I'll leave other comments on here.

galenlynch commented on src/dspbase.jl:
```diff
@@ -565,139 +570,123 @@ function unsafe_conv_kern_os!(out,
-function conv(u::AbstractArray{<:Number, N},
-              v::AbstractArray{<:BLAS.BlasFloat, N}) where N
-    conv(float(u), v)
+function conv(A::AbstractArray...)
```
`conv` requires at least two inputs: I think it would be cleaner to define all of the public facing `conv` methods to be like `conv(a, b, cs...)` to ensure that people can't call it with zero or one input.

galenlynch commented on src/dspbase.jl:
```diff
@@ -565,139 +570,123 @@ function unsafe_conv_kern_os!(out,
-              v::AbstractArray{<:Number, N}) where N
-    conv(u, float(v))
+# TODO: not the most concise notation, make base julia issue?
+conv(A::Union{AbstractArray{Complex{<:Number}, N},
```
`Complex{<:Number}` is a `Number` so I'm not sure why you need this union?

galenlynch commented on src/dspbase.jl:
```diff
@@ -565,139 +570,123 @@ function unsafe_conv_kern_os!(out,
+conv(A::AbstractArray{<:Integer}...) = round.(Int, conv([float(a) for a in A]...))
+function conv(A::AbstractArray{<:BLAS.BlasFloat, N}...) where N
+    sizes = [size(a) for a in A]
+    order = reverse(sortperm([prod(s) for s in sizes]))
```
I think this can be accomplished by a single call to `sortperm`, e.g. `sortperm(prod.(sizes), rev=true)`

HDictus replied:
ok, although `sortperm` doesn't work on tuples so I need to convert the ouput of `prod.` somehow. `convert` doesn't seem to do the trick so I think I'll have to keep using a comprehension here.

galenlynch replied:
We should decide on whether or not to sort the inputs before addressing this.

galenlynch commented on src/dspbase.jl:
```diff
@@ -565,139 +570,123 @@ function unsafe_conv_kern_os!(out,
+                    conv(promote(A...)...)
+conv(A::AbstractArray{<:Integer}...) = round.(Int, conv([float(a) for a in A]...))
+function conv(A::AbstractArray{<:BLAS.BlasFloat, N}...) where N
+    sizes = [size(a) for a in A]
```
I personally find comprehensions to be wordy, and I think they are often the source of type instability. This could easily be written as `sizes = size.(A)`.

galenlynch commented on src/dspbase.jl:
```diff
@@ -472,7 +475,9 @@ function unsafe_conv_kern_os!(out,
     # Transform the smaller filter
     _zeropad!(tdbuff, v)
-    filter_fd = os_filter_transform!(tdbuff, p)
+    filter_fd = os_filter_transform!((tdbuff,
```
Maybe we should just dispatch the whole filter transformation, including the zeropadding. In the case where there are only two inputs, then it makes sense to reuse the pad buffer and fft plan because we can expect the pad buffer to be correctly sized. In the case of three or more arrays to convolve, then we will save time by not using the pad buffer because (presumably) the trailing arrays will be missing at least a dimension compared to the first.

HDictus replied:
Ah, I had assumed they needed the same dimensions. I tried without the padding - just using fft or rfft, the time to execute the dsp.jl tests went from 71 seconds to 77. Slight slowdown, so I think it's best to revert 
edit: ran a few more trials and not padding is consistently slightly slower

galenlynch commented on src/dspbase.jl:
```diff
@@ -284,12 +284,14 @@ end
-@inline function os_filter_transform!(buff::AbstractArray{<:Complex}, p!)
-    copy(p! * buff) # p operates in place on buff
+@inline function os_filter_transform!(A::NTuple{<:Any, AbstractArray{<:Complex}}, p!)
+    fA = (copy(p! * a) for a in A) # p operates in place on buff
```
It seems like a shame to copy each transformed array, when you could probably get away with copying far fewer. ATM, you have already expanded the inputs to all be the same size, so building up the product incrementally seems better. If you instead don't pad them to be the same size (which I think might be faster), then we'd have to play around to see what the fastest method would be.

HDictus replied:
I'll get round to this soon

HDictus replied:
What was the reason initially for copying, by the way?

galenlynch replied:
I was trying to minimize the amount of memory allocated by the convolution routine, and did so be creating a zero-padded time domain buffer and transformation result Fourier domain buffer that I could re-use a bunch of times. Because we need to transform the filter only once, and subsequently use it to convolve each block, we need to keep the transformed filter in a separate piece of memory so it's not overwritten.

For real inputs, the Fourier transform needs a separate array to write the results to, or it creates a new array for you. In that case, we can just apply the transformation to the zero-padded filter to get a new array with the transformed filter. This new array does not share memory with either the time domain nor the fourier domain buffers. 

However, when working with complex inputs, the situation is a little different. First of all, there is only one "working" buffer instead of two, because both the time and Fourier domain are complex. This means we can store both the zero-padded signal and the transformed signal in the same buffer. Additionally, the complex Fourier transform operates in-place, which means the transform result is stored in the input array. So the filter is first put into the working buffer and zero padded, then it's transformed in that same buffer, but because we'll overwrite the working buffer's contents later with data from each block, we need to copy the transformed filter out of the working buffer into a new piece of memory so we can hang onto it.

galenlynch reviewed (changes requested):
Sorry that this review is taking a while. Instead of working on it until completion, I'll just do it piecemeal. At the moment this PR isn't type stable, which is a non-starter for me. Sorting the inputs by size seems like a big source of type instability that I don't know how to get around. Maybe we can just ask users to put the biggest input first, and throw a warning (or not?) if they don't? I have some other suggestions about the interface etc., and will delve into this more over the coming days.

HDictus commented:
> Sorry that this review is taking a while. Instead of working on it until completion, I'll just do it piecemeal. At the moment this PR isn't type stable, which is a non-starter for me. Sorting the inputs by size seems like a big source of type instability that I don't know how to get around. Maybe we can just ask users to put the biggest input first, and throw a warning (or not?) if they don't? I have some other suggestions about the interface etc., and will delve into this more over the coming days.

In any case, removing the sort alone isn't enough to solve the type-stability problem (I just tried that), and I'm not convinced it is a cause of it (after all, there was no type instability on the overlap-save branch when swapping the order of two inputs depending on which is larger. Output type of the subsequently called functions is independent of the order of the arguments). I'll comment out the sorting for now, and once the type instability is solved we can reenable it to see if it is still type-stable at that point.

galenlynch commented:
It might not be the only cause of it, and maybe it won't lead to type instability in the output, but changing the order of the input tuple seems inherently type unstable to me. 

First, assume that the inputs are not all the same type, maybe one input is a view, or maybe one has somehow been transformed and therefore not a standard array. In the base case with two inputs (as in the overlap-save branch), an `if` statement is used to select between the two orderings, so the compiler sees the two options for re-arranging the inputs, and both paths are completely type inferable, so no problem. For vararg inputs, we will just combine all the inputs into a tuple of arrays, with differing element types (which is the only way I can think of doing it). Different orderings of that tuple will result in different tuple types. If we then re-order the input tuple using an indexing vector that is determined at run time (e.g. the output of `sortperm`) and therefore is impossible for the compiler predict, then the type of the resulting re-ordered tuple seems impossible to infer, because each ordering will produce a different type.

As far as I understand, vector indexing into tuples shouldn't be type inferable.

HDictus committed:
some of the review advice...

public conv needs 2 or more arguments
redundant union removed
permsort with rev=true
comprehension replaced with elementwise operation

HDictus committed:
comment out sorting for now

(fix minor issue)

HDictus committed:
dispatch one and more to different filter transforms

copy only out array for complex

HDictus commented:
> It might not be the only cause of it, and maybe it won't lead to type instability in the output, but changing the order of the input tuple seems inherently type unstable to me.
> 
> First, assume that the inputs are not all the same type, maybe one input is a view, or maybe one has somehow been transformed and therefore not a standard array. In the base case with two inputs (as in the overlap-save branch), an `if` statement is used to select between the two orderings, so the compiler sees the two options for re-arranging the inputs, and both paths are completely type inferable, so no problem. For vararg inputs, we will just combine all the inputs into a tuple of arrays, with differing element types (which is the only way I can think of doing it). Different orderings of that tuple will result in different tuple types. If we then re-order the input tuple using an indexing vector that is determined at run time (e.g. the output of `sortperm`) and therefore is impossible for the compiler predict, then the type of the resulting re-ordered tuple seems impossible to infer, because each ordering will produce a different type.
> 
> As far as I understand, vector indexing into tuples shouldn't be type inferable.

Is passing the @inferred tests in dsp.jl.xcorr testset sufficient to guarantee type-stability, or should I make some more tests? Without the sorting those indeed pass.

HDictus committed:
report time for each test file

HDictus commented:
Ok, I've been playing around with `@elapsed` and some branches  where I impemented your suggestions. There doesn't seem to be a major change in the time taken to run tests/dsp.jl, but I'm now adding in some big convolutions that should hopefully make any difference more pronounced.

I also found out that part of the reason that the test comparing the time to execute a naive n-arg convolution against the implemented one fails in some cases on each branch is becasue the naive one is performed just after the implemented, and that stops when I switch them. I guess for small convolutions relevant parts must still be sitting in the cache or something when the naive conv is run.

The third thing I've noticed is that when I stick to overlap-save-conv and just implement a method conv(A, B, C...) = conv(conv(A, B), C...) runs through the tests faster than any of the other versions. I hope to see that adding some big convolutions shows that this is only the case when the arrays are small.

HDictus committed:
add some larger convolutions to test

also fix issue that test revealed

HDictus committed:
@elapsed comparison turns out to be meaningless due to caching

HDictus commented:
It turns out that there were some cases that weren't getting covered by my tests, so now I'll have to fix the versions where I had changed things

HDictus commented on test/runtests.jl:
```diff
@@ -10,5 +10,8 @@ testfiles = [ "dsp.jl", "util.jl", "windows.jl", "filter_conversion.jl",
 for testfile in testfiles
-    eval(:(@testset $testfile begin include($testfile) end))
+    @testset "$testfile" begin
+        time = @elapsed eval(:(include($testfile)))
```
When I did this I didn't consider whether compilation would be included in this `@elapsed` check. Do you know whether that is the case? If so, any idea how I can compile beforehand without running the tests, such that I only time the tests.

HDictus committed:
change mul!. for .*= as it is not defined for these types

galenlynch commented:
I don't know if I would put benchmarks in the test script.

Benchmarking correctly is tricky, as it seems you're discovering. I would
use BenchmarkTools.jl (https://github.com/JuliaCI/BenchmarkTools.jl) to do
it.

On Wed, Jul 3, 2019, 4:16 AM HDictus <notifications@github.com> wrote:

> *@HDictus* commented on this pull request.
> ------------------------------
>
> In test/runtests.jl
> <https://github.com/galenlynch/DSP.jl/pull/2#discussion_r299830053>:
>
> > @@ -10,5 +10,8 @@ testfiles = [ "dsp.jl", "util.jl", "windows.jl", "filter_conversion.jl",
>  seed!(1776)
>
>  for testfile in testfiles
> -    eval(:(@testset $testfile begin include($testfile) end))
> +    @testset "$testfile" begin
> +        time = @elapsed eval(:(include($testfile)))
>
> When I did this I didn't consider whether compilation would be included in
> this @Elapsed <https://github.com/Elapsed> check. Do you know whether
> that is the case? If so, any idea how I can compile beforehand without
> running the tests, such that I only time the tests.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/galenlynch/DSP.jl/pull/2?email_source=notifications&email_token=ABUDZTZTFF33P5FHJJMVFF3P5RN6XA5CNFSM4HNZ7RIKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5LH5ZA#pullrequestreview-257326820>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABUDZTZVOP4PE7V4JQFI6J3P5RN6XANCNFSM4HNZ7RIA>
> .
>

HDictus commented:
Hello again! I have some time again, so I took your advice on benchmarking.

I used the following benchmarks:
`
using BenchmarkTools
using DSP: conv

int1D = @benchmark conv(reverse(A)...) setup=(A=(rand(Int8, (100,)),
                                                 rand(Int8, (1000,)),
                                                 rand(Int8, (123456,))))
@show(int1D)

float1D = @benchmark conv(reverse(A)...) setup=(A=(rand(Float64, (100,)),
                                                   rand(Float64, (1000,)),
                                                   rand(Float64, (123456,))))
@show(float1D)

complex1D = @benchmark conv(reverse(A)...) setup=(
    A=(complex.(rand(Float64, (100,))),
       complex.(rand(Float64, (1000,))),
       complex.(rand(Float64, (123456,)))))
@show(complex1D)

many1D = @benchmark conv(reverse(A)...) setup=(
    A=(complex.(rand(Float64, (100,))),
       rand(Float32, (100,)),
       rand(Int8, (123456,))))
@show(many1D)

floatND = @benchmark conv(reverse(A)...) setup=(A=(rand(Float64, (100, 10)),
                                                   rand(Float64, (1, 100, 20)),
                                                   rand(Float64, (123, 10, 2))))
@show(floatND)

manyND = @benchmark conv(reverse(A)...) setup=(A=(
    complex.(rand(Float64, (100, 10))),
    rand(Int8, (1, 100, 20)),
    rand(Float64, (123, 10, 2))))

@show(manyND)
`

for the branch as is:
`
int1D = Trial(1.705 ms)
float1D = Trial(1.455 ms)
complex1D = Trial(3.058 ms)
many1D = Trial(1.827 ms)
floatND = Trial(21.688 ms)
manyND = Trial(39.617 ms)
`

the naive version:
`     
int1D = Trial(2.458 ms)
float1D = Trial(1.965 ms)
complex1D = Trial(4.108 ms)
many1D = Trial(2.499 ms)
floatND = Trial(580.800 ms)
manyND = Trial(600.964 ms)
`

copying fewer arrays, as you suggested on june 4th
`
int1D = Trial(1.674 ms)
float1D = Trial(1.434 ms)
complex1D = Trial(2.992 ms)
many1D = Trial(1.692 ms)
floatND = Trial(21.600 ms)
manyND = Trial(38.836 ms)
`
The branch where I was using an unplanned fft in order to avoid the extra padding somehow raised an error, so I'll have to have a closer look at that. (edit: it also fails tests, so I suppose I never got it working. for some array shapes the ffts cannot be multplied together, are you sure the padding is unnecessary?)

So it seems that it is substantially faster than the naive approach! as the last results are the fastest I'll merge those changes in.

galenlynch commented:
Hey! Sorry that I've been absent: I was trying to finish up my degree and defend. Now that I'm done I should have some time to look at this in the near future. I hope you're still interested in working on this!

HDictus commented:
Hey @galenlynch Congratulations on your degree and defense! 
I'm still interested! I've just been busy during October, moving apartments + lots of work. 
how about you merge your branch first, and then I'll rebase to master and re-make the PR on the main repo
